source/beamline/AMPseudoMotorControl.cpp

```diff
@@ -0,0 +1,218 @@
+#include "AMPseudoMotorControl.h"
+#include "util/AMErrorMonitor.h"
+
+AMPseudoMotorControl::AMPseudoMotorControl(const QString &name, const QString &units, QObject *parent, const QString &description) :
+	AMControl(name, units, parent, description)
+{
+	// Initialize local variables.
+
+	connected_ = false;
+	value_ = 0;
+	setpoint_ = 0;
+	moveInProgress_ = false;
+	isMoving_ = false;
+	minimumValue_ = 0;
+	maximumValue_ = 0;
+
+	startedMapper_ = new QSignalMapper(this);
+	connect( startedMapper_, SIGNAL(mapped(QObject*)), this, SLOT(onMoveStarted(QObject*)) );
+
+	cancelledMapper_ = new QSignalMapper(this);
+	connect( cancelledMapper_, SIGNAL(mapped(QObject*)), this, SLOT(onMoveCancelled(QObject*)) );
+
+	failedMapper_ = new QSignalMapper(this);
+	connect( failedMapper_, SIGNAL(mapped(QObject*)), this, SLOT(onMoveFailed(QObject*)) );
+
+	succeededMapper_ = new QSignalMapper(this);
+	connect( succeededMapper_, SIGNAL(mapped(QObject*)), this, SLOT(onMoveSucceeded(QObject*)) );
+}
+
+AMPseudoMotorControl::~AMPseudoMotorControl()
+{
+
+}
+
+AMControl::FailureExplanation AMPseudoMotorControl::move(double setpoint)
+{
+	if (!isConnected()) {
+		AMErrorMon::alert(this, AMPSEUDOMOTORCONTROL_NOT_CONNECTED, QString("Failed to move %1: control is not connected.").arg(name()));
+		return AMControl::NotConnectedFailure;
+	}
+
+	if (isMoving()) {
+		AMErrorMon::alert(this, AMPSEUDOMOTORCONTROL_ALREADY_MOVING, QString("Failed to move %1: control is already moving.").arg(name()));
+		return AMControl::AlreadyMovingFailure;
+	}
+
+	if (!validSetpoint(setpoint)) {
+		AMErrorMon::alert(this, AMPSEUDOMOTORCONTROL_INVALID_SETPOINT, QString("Failed to move %1: provided setpoint is invalid.").arg(name()));
+		return AMControl::LimitFailure;
+	}
+
+	// Update the setpoint.
+
+	setSetpoint(setpoint);
+
+	// Create new move action.
+
+	AMAction3 *moveAction = createMoveAction(setpoint);
+
+	if (!moveAction) {
+		AMErrorMon::alert(this, AMPSEUDOMOTORCONTROL_INVALID_MOVE_ACTION, QString("Failed to move %1: invalid move action generated.").arg(name()));
+		return AMControl::LimitFailure;
+	}
+
+	// Create move action signal mappings.
+
+	startedMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(started()), startedMapper_, SLOT(map()) );
+
+	cancelledMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(cancelled()), cancelledMapper_, SLOT(map()) );
+
+	failedMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(failed()), failedMapper_, SLOT(map()) );
+
+	succeededMapper_->setMapping(moveAction, moveAction);
+	connect( moveAction, SIGNAL(succeeded()), succeededMapper_, SLOT(map()) );
+
+	// Run action.
+
+	moveAction->start();
+
+	return AMControl::NoFailure;
+}
+
+bool AMPseudoMotorControl::stop()
+{
+	bool result = true;
+
+	foreach (AMControl *child, children_) {
+		if (child)
```

Comment:
To be thorough, you may want to have this if-statement incorporate a childs ability to stop (i.e.: ```if (child && child->canStop())```).  Food for thought.  That may open up other issues, but it probably makes sense to think about them now.